### PR TITLE
Allow lists of SkyCoord/frames/representations in SkyCoord init

### PR DESCRIPTION
**Pull request by eteq**
This PR is a different take on #2702 (which it replaces, and has been in limbo since just before 0.4).  It does the main point of that, allowing usage like:
```
>>> sc = SkyCoord(ra=[1,2,3]*u.deg,dec=[4,5,6]*u.deg)
>>> sc0 = sc[0]
>>> sc2 = sc[2]
>>> SkyCoord([sc0, sc2])
```
However, it's more restricted than #2702, which allowed various mixes of types of frames and converted as needed to match the first one.  Based on @taldcroft's last comments in #2702, it no longer does that, and always raises an error if the input frames are not all consistent.

This is also different from #2702 in that it is a no-op if the input list does not include SkyCoords, frame objects, or representations.  So unlike #2702, it should have no effects outside that case, which previously was not allowed.  (The big block of changes in the diff is mostly just indenting in an else statement - all of that is unchanged aside from the indednt)  

I'll leave it to @astrofrog to decide if this should be aimed for v1.0, 1.0.1, or 1.1 .

### CODE REVIEW

**Comment by eteq** [comment]
Oh, and if it's not clear, this closes #2702 when it gets merged. If we do go with this, I'll create an issue to (in a later version) add a `create_from_heterogeneous` `SkyCoord` class method, as discussed in #2702

File: astropy/coordinates/sky_coordinate.py [hunk 1]
```diff
@@ -1147,14 +1150,19 @@ def _get_units(args, kwargs):
     return units
 
 
-def _parse_coordinate_arg(coords, frame, units):
+def _parse_coordinate_arg(coords, frame, units, scinit_kwargs):
```
Comment (embray) [inline astropy/coordinates/sky_coordinate.py]: I feel like just `init_kwargs` or even just `kwargs` or `options` would be fine names for this.

File: astropy/coordinates/sky_coordinate.py [hunk 2]
```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
...
-        # Now transpose vals to get [(v1_0 .. v1_N), (v2_0 .. v2_N), (v3_0 .. v3_N)]
-        # (ok since we know it is exactly rectangular).  (Note: can't just use zip(*values)
-        # because Longitude et al distinguishes list from tuple so [a1, a2, ..] is needed
-        # while (a1, a2, ..) doesn't work.
-        values = [list(x) for x in zip(*vals)]
-
-        if is_scalar:
-            values = [x[0] for x in values]
-
+        if hascoordobjs:
```
Comment (embray) [inline astropy/coordinates/sky_coordinate.py]: I feel like these different paths deserve separate subroutines.  `_init_from_foo`, `_init_from_bar`, etc.

File: astropy/coordinates/sky_coordinate.py (hunk 1 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: I think ``kwargs`` isn't good because it makes it seem like it's kwargs to this function, rather than the `SkyCoord`.  But you're right `init_kwargs` makes more sense so I'll do that.

File: astropy/coordinates/sky_coordinate.py (hunk 2 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: @embray - I agree in principal, but when I tried to do this, it ended up quite awkward because of the need to pass around a bunch of variables used inside like ``valid_kwargs``, ``frame_attr_names``, ``repr_attr_names``, etc.

I actually think that a lot of this parsing structure needs to be re-factored - even without this PR the flow of the code is getting rather confusing.  But I think that's a much larger scope than this PR, so I'd rather have this in and then later on (probably or 1.1)  do a larger-scale reorg.

**Comment by eteq** [comment]
rebased, and one of @embray's comments addressed (the other I'm suggesting we defer - see above)

File: astropy/coordinates/sky_coordinate.py [hunk 3]
```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
...
+            # and frame name
+            if len(set((sc.frame.name for sc in scs if hasattr(sc, 'frame')))) > 1:
+                frames = [sc.frame.name if hasattr(sc, 'frame') else None for sc in scs]
+                raise ValueError("List of inputs have different frames: {0}".format(frames))
+            for fattrnm in FRAME_ATTR_NAMES_SET():
+                vals = [getattr(sc, fattrnm) for sc in scs]
+                for val in vals[1:]:
+                    if val != vals[0]:
+                        raise ValueError("List of inputs don't give consistent "
+                                         "frame attribute {0}: {1}".format(fattrnm, vals))
```
Comment (astrofrog) [inline astropy/coordinates/sky_coordinate.py]: A comment for another time probably, but I feel that we want to have a method on frames that allows them to be compared to other frames (excluding data), so e.g.

```
frame1.is_equivalent_to(frame2)
```

or something like this. Doesn't have to be done here though.

**Comment by astrofrog** [comment]
:+1:

**Comment by astrofrog** [comment]
(and 1.0 is fine)

File: astropy/coordinates/sky_coordinate.py (hunk 2 above)
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: Would a partial refactor be possible using inline functions that are defined after `valid_kwargs` et al. get set?

File: astropy/coordinates/sky_coordinate.py [hunk 4]
```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
...
     elif isinstance(coords, (collections.Sequence, np.ndarray)):
-        # Handles generic list-like input.
+        # Handles list-like input.
 
-        # First turn into a list of lists like [[v1_0, v2_0, v3_0], ... [v1_N, v2_N, v3_N]]
+        # First turn into a list of lists like [[v1_0, v2_0, v3_0], ... [v1_N, v2_N, v3_N]],
+        # and check if there are any coordinate objects mixed in
         vals = []
-        for ii, coord in enumerate(coords):
+        hascoordobjs = False
```
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: What about reducing the complexity a bit with
```
coord_classes = (SkyCoord, BaseCoordinateFrame, BaseRepresentation)
if any(isinstance(coord, coord_classes) for coord in coords):
    # This parsing path is used ...
else:
    vals = []
    for coord in cords:
        if isinstance(coord, six.string_types):
            ...
```
You do pay a penalty of sometimes having the top-level loop run twice, but I think in practice the overhead for just the looping is quite small compared to everything else.   The benefit is that the code logic is more distinct.  And if it is possible to refactor the two code paths as I suggested above then things get even more clear.

I definitely agree this part of the code can use some TLC.  BTW, worth noting #2920 which is fiddling in the same function.  Will likely conflict but not in a difficult way.

File: astropy/coordinates/sky_coordinate.py [hunk 5]
```diff
@@ -227,7 +230,7 @@ def _parse_inputs(self, args, kwargs):
                 # coord_kwargs will contain keys like 'ra', 'dec', 'distance'
                 # along with any frame attributes like equinox or obstime which
                 # were explicitly specified in the coordinate object (i.e. non-default).
-                coord_kwargs = _parse_coordinate_arg(args[0], frame, units)
+                coord_kwargs = _parse_coordinate_arg(args[0], frame, units, kwargs)
```
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: Shouldn't this be `valid_kwargs`?  At this point `kwargs` has been verified to be empty as the result of popping off the valid kwargs.

File: astropy/coordinates/sky_coordinate.py [hunk 6]
```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
...
-
+        if hascoordobjs:
+            # this parsing path is used when there are coordinate-like objects
+            # in the list - instead of creating lists of values, we create
+            # SkyCoords from the list elements and then combine them.
+            scs = [SkyCoord(coord, **init_kwargs) for coord in coords]
+
+            # now check that they're all self-consistent in their frame attributes
+            # and frame name
+            if len(set((sc.frame.name for sc in scs if hasattr(sc, 'frame')))) > 1:
```
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: Is there some way that `SkyCoord` object can *not* have a `frame` attribute?  Unless I'm missing some code path the `_get_frame` function is guaranteed to return an instance of a frame class.

File: astropy/coordinates/sky_coordinate.py [hunk 7]
```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
...
+                frames = [sc.frame.name if hasattr(sc, 'frame') else None for sc in scs]
```
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: Nit, but if you have `frames = set(sc.frame.name for sc in scs)` above the `if` statement you can just do `if len(frames) > 1:`  and avoid repeating code to generate the exception message.

File: astropy/coordinates/sky_coordinate.py [hunk 8]
```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
...
+                    if val != vals[0]:
+                        raise ValueError("List of inputs don't give consistent "
+                                         "frame attribute {0}: {1}".format(fattrnm, vals))
+
+                valid_kwargs[fattrnm] = getattr(scs[0], fattrnm)
+
+            # Now combine the values, to be used below
+            values = []
+            for aname in frame_attr_names:
+                values.append([])
```
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: This will end up generating individual `Quantity` objects for every element of each `sc`, so if could really blow up for large input `SkyCoord` objects.  Here's an idea, completely untested:
```
for aname in frame_attr_names:
    attr_vals = []
    for sc in scs:
        attr_val = getattr(sc, aname)
        attr_vals.append(attr_val.reshape(1,) if sc.isscalar else attr_val)
    concat_vals = np.concatenate(attr_vals)
    # Hack because np.concatenate doesn't fully work with Quantity
    if isinstance(concat_vals, Quantity):
        concat_vals._unit = attr_val.unit
    values.append(concat_vals)
```

**Comment by taldcroft** [comment]
This improvement effectively provides a great way to concatenate a list of coordinate objects.  This might not be immediately obvious to users so could be documented or else spun into a new high-level convenience function `coordinates.concatenate`.

**Comment by taldcroft** [comment]
@eteq - done with my review.

**Comment by eteq** [comment]
I'm working through these now (although my guess is I won't get it done until tomorrow). 

@taldcroft re your final comment: You're definitely right I should add an example of this to the docs.  But I was thinking something like ``coordinates.concatenate`` would be a better place for the "inhomogenous" joiner that we were discussing in #2702 (although not until 1.1).  So the way it would work in this case is that it would always fail unless the frames completely matched if you do ``SkyCoord([...])``, but ``coordinates.concatenate([...])`` would instead try to convert to all the same frame (the exact way of determining that can be decided later).  Sound good?

**Comment by taldcroft** [comment]
@eteq - yes, sounds good.  The thing is that if there is going to be a `coordinates.concatenate` in the future, we could put it in *now* as a more limited version, and then take away restrictions in the future.  There's nothing wrong with having a concatenate that is not (yet) all-powerful.

**Comment by eteq** [comment]
@taldcroft - you mean just something like

```
def concatenate(*args,**kwargs):
    if not isiterable(args[0]):
        raise SomeError
    return SkyCoord(*args, **kwargs)
```

As a placeholder? We could do that, but that locks us to an API that looks like SkyCoord even for an inhomogeneous case down the road. I think I'm OK with that if you are, but I figured I should check that's what you mean...

**Comment by taldcroft** [comment]
I was thinking of something even simpler, that I think would actually cover some useful cases right now:
```
def concatenate(coords):
    if not isiterable(coords):
        raise SomeError
    return SkyCoord(coords)
```
Yes, it's limited, but the utility would be provided by 1) discoverability (people google/tab-complete for concatenate) and 2) the docstring which specifically discusses the concatenation use case.

By making the API very limited in this version there is no worry about API locking, since *any* concatenate method down the line will start with a positional arg of list of the things to concat.

**Comment by taldcroft** [comment]
BTW, the reason I would go for the simplest possible API now is that by allowing the pass-thru args and kwargs you open up a lot more ways that things can go wrong and more uncertainty in the API.  If/when users actually complain that they need more flexibility then this can guide development.

**Comment by eteq** [comment]
Gotcha - that makes sense.

File: astropy/coordinates/sky_coordinate.py (hunk 5 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: but the ``coord_kwargs`` have *not* yet been checked for conflicts - that's what the process of copying from ``coord_kwargs`` to ``valid_kwargs`` below.  If I just add something like
```
valid_kwargs.update(coord_kwargs)
coord_kwargs = {}
```
below this, some of the tests like
```
sc = SkyCoord(1, 2, frame='icrs', unit='deg', equinox='J1999', obstime='J2001')
with pytest.raises(ValueError) as err:
    SkyCoord(sc, equinox='J1999', obstime='J2002')
```
now fail, because the second `SkyCoord` is created just fine (which we *don't* want)

File: astropy/coordinates/sky_coordinate.py (hunk 6 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: good catch- this was from an earlier implementation where they weren't always converted to `SkyCoord` already, but now they are, so no need for this.

File: astropy/coordinates/sky_coordinate.py (hunk 7 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: I actually intentionally avoided that because I did some performance tests that suggested this approach is significantly faster. But I now find the opposite on a different platform, implying that was premature optimization... so I'll switch it to your suggestion as that is a bit clearer to read.

File: astropy/coordinates/sky_coordinate.py (hunk 8 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: needed some other tweaking, but basically that does the job - thanks!

File: astropy/coordinates/sky_coordinate.py (hunk 4 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: Good idea, that's a lot clearer - will do!

File: astropy/coordinates/sky_coordinate.py (hunk 2 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: It's certainly possible, but it has a performance penalty (although probably rather small), and at least to me, any but the simplest inline functions tend to make the code harder to follow.  So I think it's clearer this way (but of course that's subjective...).

**Comment by eteq** [comment]
Alright, I think everything here has been addressed, so if the tests pass it should be all set!

File: astropy/coordinates/sky_coordinate.py (hunk 3 above)
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: that's a good idea - will try to make a separate PR to implement this (or if it starts getting too complicated for the time remaining, an issue)

File: astropy/coordinates/sky_coordinate.py (hunk 5 above)
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: @eteq- I'm not sure we're on the same page here.  I'm looking at the code and on line 228 there is the statement:
```
if kwargs:
    raise ValueError(...)
```
Then on line 240 there is the call `coord_kwargs = _parse_coordinate_arg(args[0], frame, units, kwargs)`, which must be the same as `coord_kwargs = _parse_coordinate_arg(args[0], frame, units, {})`.  

I haven't traced through the logic but I was suggesting `coord_kwargs = _parse_coordinate_arg(args[0], frame, units, valid_kwargs)`.   My guess is that it would allow something like:
```
sc = SkyCoord('1d', '2d')
sc2 = SkyCoord([sc, sc], obstime='J2001.12')
```
where `obstime` becomes a `valid_kwarg`.

**Comment by astrofrog** [comment]
@eteq - thanks!

File: astropy/coordinates/sky_coordinate.py (hunk 5 above)
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: @astrofrog @eteq - I think this comment is still unresolved, though it's not a showstopper.  I've reproduced the failure that @eteq showed above so I understand a fix is not so simple, but in the current code (now in master) there is the following inconsistency:
```
In [1]: from astropy.coordinates import *

In [2]: sc = SkyCoord('1d', '2d', frame='fk5')

In [3]: sc2 = SkyCoord(sc, obstime='J2002')

In [4]: sc2.obstime
Out[4]: 'J2002'

In [5]: sc2 = SkyCoord([sc, sc], obstime='J2002')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-753b93d8b73e> in <module>()
----> 1 sc2 = SkyCoord([sc, sc], obstime='J2002')

/Users/aldcroft/git/astropy/astropy/coordinates/sky_coordinate.pyc in __init__(self, *args, **kwargs)
    146         # creating the internal self._sky_coord_frame object
    147         args = list(args)  # Make it mutable
--> 148         kwargs = self._parse_inputs(args, kwargs)
    149 
    150         # Set internal versions of object state attributes

/Users/aldcroft/git/astropy/astropy/coordinates/sky_coordinate.pyc in _parse_inputs(self, args, kwargs)
    262                     raise ValueError("Coordinate attribute '{0}'={1!r} conflicts with "
    263                                      "keyword argument '{0}'={2!r}"
--> 264                                      .format(attr, coord_value, valid_kwargs[attr]))
    265                 valid_kwargs[attr] = coord_value
    266 

ValueError: Coordinate attribute 'obstime'=None conflicts with keyword argument 'obstime'='J2002'
```

In addition as I mentioned the new `init_kwargs` argument of `_parse_coordinate_arg` is essentially orphaned because the input `kwargs` is guaranteed to be `{}` all the time.  I actually changed this in the code to be `coord_kwargs = _parse_coordinate_arg(args[0], frame, units, {})` and all tests passed.
